[MoveOnlyAddressChecker] Maximize lifetimes.

nate-chandler opened the pull request:
Previously, the checker inserted destroys after each last use.  Here, extend the lifetimes of fields as far as possible within their original (unchecked) limits.

rdar://99681073

nate-chandler committed:
[FieldSensitivePL] NFC: Implemented print.

The members were declared but undefined.

nate-chandler committed:
[MoveOnlyAddressChecker] NFC: Promoted assertion.

Dumped more info and called llvm_unreachable on bad state.

nate-chandler committed:
[MoveOnlyAddressChecker] Maximize lifetimes.

Previously, the checker inserted destroys after each last use.  Here,
extend the lifetimes of fields as far as possible within their original
(unchecked) limits.

rdar://99681073

atrick commented on lib/SILOptimizer/Mandatory/MoveOnlyAddressCheckerUtils.cpp:
```diff
@@ -631,9 +631,17 @@ void swift::siloptimizer::searchForCandidateAddressMarkMustChecks(
 
+using ConsumingBlocksCollection = SmallPtrSetVector<SILBasicBlock *, 8>;
+using ConsumingBlocks =
+    llvm::SmallMapVector<TypeTreeLeafTypeRange, ConsumingBlocksCollection, 8>;
```
Can this be a single `SmallMapVector<SILBasicBlock*, SmallBitSet>` to map each ConsumingBlock to the fields that are consumed? When you query a field's consumes, you would just filter the result.

That would be consistent with how we represent per-block liveness state.
There are an exponential number of ranges. Whereas blocks are linear, and in practice we have a very small number, and field assignments often occur in the same block.

nate-chandler replied:
Switched over to this storage.

atrick reviewed (commented):
Awesome! Just a few areas of confusion...

nate-chandler commented on lib/SILOptimizer/Mandatory/MoveOnlyAddressCheckerUtils.cpp:
```diff
@@ -2634,6 +2687,236 @@ void MoveOnlyAddressCheckerPImpl::rewriteUses(
+        continue;
+      // Add "the instruction(s) before the terminator" of the predecessor to
+      // liveness.
+      addPreviousInstructionToLiveness(predecessor->getTerminator(), element,
```
This is actually straight from the original version in CanonicalizeOSSALifetime (though the addPreviousInstructionToLiveness was extracted into a separate function since it's needed for destroys as well).  Liveness is extended "to the end" of these blocks.  The terminators themselves can't be added to liveness or else destroys would be created in their successors; instead, their previous instructions are added.  If there's no previous instruction in terminator's block, the terminators of its predecessors are added.  This is what we want because it still results in liveness extended to just before the terminator in question and the destroy being created just before it.

nate-chandler commented on lib/SILOptimizer/Mandatory/MoveOnlyAddressCheckerUtils.cpp:
```diff
@@ -2634,6 +2687,236 @@ void MoveOnlyAddressCheckerPImpl::rewriteUses(
+      auto pair = liveness.isInterestingUser(instruction);
+      if (pair.first != FieldSensitivePrunedLiveness::IsInterestingUser::
+                            LifetimeEndingUse ||
+          !pair.second.has_value() || !pair.second->contains(element)) {
```
Thanks!  Fixed.

nate-chandler commented on lib/SILOptimizer/Mandatory/MoveOnlyAddressCheckerUtils.cpp:
```diff
@@ -2634,6 +2687,236 @@ void MoveOnlyAddressCheckerPImpl::rewriteUses(
+/// - In particular, in regions of originalLiveBlocks which have no boundary
+///   with consumedAtExitBlocks, liveness should be extended to its original
+///   extent.
+///   [Extend liveness down to the boundary between green blocks and uncolored.]
```
Added that.

atrick reviewed (approved):
Nice!

nate-chandler committed:
[TypeLowering] Move-only types are lexical.

Vars of such types should be given lexical `alloc_stack`s by
`AllocBoxToStack` which requires that the `alloc_box` insts formed for
them have an associated borrow scope which in turn requires that type
lowering for move only structs and enums have their lexical bits set.

rdar://110901430

nate-chandler committed:
[Basic] Removed spurious req from FrozenMultiMap.

Its storage vector is intended to be of some type like
`std::vector<std::pair<Key, Optional<Value>>>`, i.e., some collection of
pairs whose `second` is an `Optional<Value>`.  So when constructing a
default instance of that pair, just construct an Optional in the None
case.

nate-chandler committed:
[SILOpt] Don't opt move-only lifetimes.

According to language rules, such lifetimes are fixed and the relative
order of their deinits is guaranteed.

rdar://110913116

nate-chandler committed:
[FieldSensitivePL] Fix vectorization.

FieldSensitivePrunedLiveness is used as a vectorization of
PrunedLiveness.  An instance of FSPL with N elements needs to be able to
represent the same states as N instances of PL.

Previously, it failed to do that in two significant ways:

(1) It attempted to save space for which elements were live by using
    a range.  This failed to account for instructions which are users of
    non-contiguous fields of an aggregate.

    apply(
      @owned (struct_element_addr %s, #S.f1),
      @owned (struct_element_addr %s, #S.f3)
    )

(2) It used a single bit to represent whether the instruction was
    consuming.  This failed to account for instructions which consumed
    some fields and borrowed others.

    apply(
      @owned (struct_element_addr %s, #S.f1),
      @guaranteed (struct_element_addr %s, #S.f2)
    )

The fix for (1) is to use a bit vector to represent which elements
are used by the instruction.  The fix for (2) is to use a second bit
vector to represent which elements are _consumed_ by the instruction.

Adapted the move-checker to use the new representation.

rdar://110909290

nate-chandler committed:
[move-only] Avoid loc from func decl.

It's always the first line of the function, so try to do better.

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
[MoveOnlyAddressChecker] Added extension flag.

Passing

```
-Xllvm -move-only-address-checker-disable-lifetime-extension=true
```

will skip the maximization of unconsumed field lifetimes.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

atrick reviewed (approved):
Thanks. I reviewed for CCC.
